src/core/src/main/java/org/apache/jmeter/samplers/SampleResult.java

```diff
@@ -791,6 +801,27 @@ public void setResponseData(final String response, final String encoding) {
      * @return the responseData value (cannot be null)
      */
     public byte[] getResponseData() {
+        if (responseData == null) {
+            return EMPTY_BA;
+        }
+        if (contentEncoding != null && responseData.length > 0) {
+            try {
+                switch (contentEncoding.toLowerCase(Locale.ROOT)) {
+                    case "gzip":
+                        return decompressGzip(responseData);
+                    case "x-gzip":
+                        return decompressGzip(responseData);
+                    case "deflate":
+                        return decompressDeflate(responseData);
+                    case "br":
+                        return decompressBrotli(responseData);
+                    default:
+                        return responseData;
+                }
+            } catch (IOException e) {
+                log.warn("Failed to decompress response data", e);
+            }
+        }
```

Comment:
Have you considered something like `interface ResponseDecoder { String decode(byte[] contents); }` or `interface Response { String toString(); byte[] toByteArray(); }`?

Then there could be something like `SampleResult#setResponse(Response)` or `SampleResult#setResponseDecoder(ResponseDecoder)`?

It just does not sound quite right to tie `SampleResult` with specific implementations of the decoding